File: 백준/Silver/2583. 영역 구하기/영역 구하기.py
import sys
from collections import deque


def bfs(x, y):
  queue = deque()
  queue.append((x, y))
  visited[x][y] = True
  result = 0
  while queue:
    x, y = queue.popleft()
    for i in range(4):
      nx = x + dx[i]
      ny = y + dy[i]
      if nx < 0 or nx >= m or ny < 0 or ny >= n:
        continue
      if visited[nx][ny]:
        continue
      if not visited[nx][ny]:
        graph[nx][ny] = graph[x][y] + 1
        result += 1
        visited[nx][ny] = True
        queue.append((nx, ny))
  return result


m, n, k = map(int, sys.stdin.readline().rstrip().split())
dx = [0, 0, 1, -1]
dy = [1, -1, 0, 0]
graph = [[0] * (n) for _ in range(m)]
visited = [[False] * (n) for _ in range(m)]

for _ in range(k):
  lx, ly, rx, ry = map(int, sys.stdin.readline().rstrip().split())
  for i in range(ly, ry - ly + ly):
    for j in range(lx, rx - lx + lx):
      graph[i][j] = 1
      visited[i][j] = True

result = 0
answer = []
for i in range(m):
  for j in range(n):
    if not visited[i][j]:
      answer.append(bfs(i, j) + 1)
      result += 1

answer.sort()
print(result)
for i in answer:
  print(i, end=' ')

File: 백준/Silver/2583. 영역 구하기/영역 구하기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
  static int m;
  static int n;
  static int k;
  static int[][] graph;
  static int[][] visited;

  static int bfs(int x, int y) {
    Queue<int[]> queue = new LinkedList<>();
    visited[x][y] = 1;
    queue.offer(new int[] { x, y });

    int[] dx = { -1, 1, 0, 0 };
    int[] dy = { 0, 0, -1, 1 };

    int area = 0;
    while (!queue.isEmpty()) {
      area++;
      int[] curr = queue.poll();
      for (int i = 0; i < 4; i++) {
        int nx = curr[0] + dx[i];
        int ny = curr[1] + dy[i];
        if (nx < 0 || nx >= m || ny < 0 || ny >= n)
          continue;
        if (graph[nx][ny] == 1 || visited[nx][ny] != -1)
          continue;
        visited[nx][ny] = visited[curr[0]][curr[1]] + 1;
        queue.offer(new int[] { nx, ny });
      }
    }
    return area;
  }
  public static void main(String[] args) throws IOException {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    StringTokenizer st = new StringTokenizer(br.readLine());
    m = Integer.parseInt(st.nextToken());
    n = Integer.parseInt(st.nextToken());
    k = Integer.parseInt(st.nextToken());
    graph = new int[m][n];
    visited = new int[m][n];
    for (int i = 0; i < m; i++) {
      Arrays.fill(visited[i], -1);
    }
    for (int cnt = 0; cnt < k; cnt++) {
      st = new StringTokenizer(br.readLine());
      int lx = Integer.parseInt(st.nextToken());
      int ly = Integer.parseInt(st.nextToken());
      int rx = Integer.parseInt(st.nextToken());
      int ry = Integer.parseInt(st.nextToken());
      for (int i = ly; i < ry; i++) {
        for (int j = lx; j < rx; j++) {
          graph[i][j] = 1;
        }
      }
    }

    int result = 0;
    ArrayList<Integer> areaArr = new ArrayList<Integer>(); 
    for (int i = 0; i < m; i++) {
      for (int j = 0; j < n; j++) {
        if (graph[i][j] == 0 && visited[i][j] == -1) {
          int area = bfs(i, j);
          areaArr.add(area);
          result++;
        }
      }
    }

    System.out.println(result);
    Collections.sort(areaArr);
    for (int i = 0; i < result; i++) {
      System.out.print(areaArr.get(i) + " ");
    }
    System.out.println();
  }
}